### PR TITLE
Eligible number of build versions

### PR DESCRIPTION
**Pull request by dimpar**
For simplicity purposes, we supported only 2 keep client release tags in a given interval. In rare cases when we need to release a hotfix it might happen that in a given interval 3 keep client versions will be eligible, just like for October 2023 interval. This PR adds support for a third release. However, the default number of tags should be set to 2, only in rare cases the number should be bumped up to 3, i.e. when 3 tags are eligible to run in a given interval.

**Important**! After the October interval rewards are calculated, the `ELIGIBLE_NUMBER_OF_TAGS` should be changed to `2`

TODO:
- [ ] `ALLOWED_UPGRADE_DELAY` will need to be adjusted, because `-m6` tag was already created but not announced. We count 3 weeks from the day of the announcement.

Also, this PR includes a cleanup of the Oct 17 2023 case. This scenario is no longer needed to be supported by these scripts.

### CODE REVIEW

File: src/scripts/tbtcv2-rewards/rewards.ts [hunk 1]
```diff
@@ -525,21 +507,16 @@ async function getAuthorization(
 function authorizationForRewardsInterval(
```
Comment (manumonti) [inline src/scripts/tbtcv2-rewards/rewards.ts:507]: Should we remove the comment above? lines 493 to 506:

https://github.com/threshold-network/merkle-distribution/blob/055f8499962eeb798f9532e4095f4590762358f6/src/scripts/tbtcv2-rewards/rewards.ts#L493-L506

**Review by manumonti — CHANGES_REQUESTED** [review]

File: src/scripts/tbtcv2-rewards/rewards.ts [hunk 2]
```diff
@@ -372,12 +372,7 @@ export async function calculateRewards() {
           // upgrade cutoff date that happens to be right before the interval
           // end date. However, it might still be eligible for rewards because
           // of the uptime requirement.
-          if (
-            !(
-              instances[i].buildVersion.includes(latestClientTag) ||
-              instances[i].buildVersion.includes(secondToLatestClientTag)
-            )
-          ) {
+          if (!eligibleClientTags.includes(instances[i].buildVersion)) {
```
Comment (manumonti) [inline src/scripts/tbtcv2-rewards/rewards.ts:375]: If I'm understanding it correctly, in the case of three versions (tags), the same cutoff date is set for the first and the second version, right?

I mean, if we have a distribution where `-m4`, `-m5` and `-m6` versions have been valid, we are setting that -m4 and -m5 are valid until cutoff date. This is, we are not checking if the node ran `-m4` version after the three-week deadline (in which case, the node should be punished).

Is it intended to simplify the application of the rules?

**Review by manumonti — CHANGES_REQUESTED** [review]

File: src/scripts/tbtcv2-rewards/rewards.ts (hunk 2 above)
Comment (dimpar) [reply src/scripts/tbtcv2-rewards/rewards.ts:375]: Correct, this is to simplify rules and not force stakers to update their nodes one after another in a short period of time.

File: src/scripts/tbtcv2-rewards/rewards.ts (hunk 1 above)
Comment (dimpar) [reply src/scripts/tbtcv2-rewards/rewards.ts:493]: Right, this should be removed. Here [e2b0a63](https://github.com/threshold-network/merkle-distribution/pull/89/commits/e2b0a63be04cb80f4d11d37c5b9a04f23a14fc85)

**Review by manumonti — APPROVED** [review]
LGTM! 🙌

**Comment by lukasz-zimnoch** [comment]
I think we should update `ALLOWED_UPGRADE_DELAY` before merging. I'll take care of it since I'm replacing @dimpar on the rewards support front.

**Comment by manumonti** [comment]
> I think we should update `ALLOWED_UPGRADE_DELAY` before merging. I'll take care of it since I'm replacing @dimpar on the rewards support front.

I think we can address this update on ALLOWED_UPGRADE_DELAY in the following PR.

**Pull request merged**